### ISSUE TITLE
Transparent text selection

### ISSUE THREAD

**Issue by workcomplete** [vedang/pdf-tools#149]
I'll preface this by saying I am not sure if this is an issue with pdf-tools specifically, vs a more general issue with tesseract OCR/poppler. I originally posted this in discussions, but I don't think many people are using this feature atm.

When I select text (i.e. click and drag mouse over a region of text) in pdf-tools, the text selection is not transparent. For pdf's that are created digitally this is fine, but for pdf's that are scanned and OCR'd with tesseract, the selected text becomes hidden behind the selection (see images below). Is there a way to change this behavior without re-OCRing the pdf with something like adobe acrobat (i.e. is it possible make the text selection transparent in pdf-tools)?

What selected text looks like in a pdf generated from a website:
![Screenshot from 2022-09-08 20-27-51](https://user-images.githubusercontent.com/77472800/189266051-fd094520-e9b9-4dcf-96b0-5611c223ae29.png)

What selected text looks like in a scanned book that has been OCRd with tesseract

![Screenshot from 2022-09-08 19-07-11](https://user-images.githubusercontent.com/77472800/189266049-874a3e57-afdd-40e7-a163-a685e24605f1.png)

OCR text can still be copied and pasted and looks fine when markup is applied.

![Screenshot from 2022-09-08 19-07-55](https://user-images.githubusercontent.com/77472800/189266050-43dfdea6-9814-4b92-8840-8c577db245ae.png)

_Originally posted by @workcomplete in https://github.com/vedang/pdf-tools/discussions/147_

If I understand this post on stack exchange, poppler has already implemented transparent text selection?

https://tex.stackexchange.com/questions/565909/invisible-text-even-when-selected-in-evince

This same issue with pdf-tools (under previous maintainer) is mentioned here

https://gitlab.freedesktop.org/poppler/poppler/-/issues/157

**Comment by orgtre** [comment]
Ok, I can reproduce this using poppler 22.08 with the `linn.pdf` attached to the poppler issue report linked above. Still, highlighted text displays correctly. In Evince the behavior is exactly the same as in pdf-tools (non-transparent when selected, but transparent when highlighted). In several other non-popper-based pdf readers the selections are transparent, which to me indicates that this is a poppler issue (despite the fact that the issue linked above is closed). Not sure if there is something we can do on the pdf-tools side.

**Comment by workcomplete** [comment]
OK well I found a work around (below) but agree that this seems to be a poppler issue of not handling `3 Tr`as it is purported to in the chosen answer in the stackexchange question above.

By following the instructions [here](https://unix.stackexchange.com/questions/171940/how-can-i-convert-a-scanned-pdf-with-ocred-text-to-one-without-ocred-text/181644#181644) up to step 2, where instead replacing `3 Tr` with `1 Tr`, I replaced `3 Tr` with `7 Tr`. This makes the text visible when selected (see image below). The remaining steps are not necessary. The downside here is that I download many books from online resources, and 80% of the time they are already OCR'd with tesseract. Needing to do this for every text is a bit painful.

![Screenshot from 2022-09-15 11-40-20](https://user-images.githubusercontent.com/77472800/190483989-96d05f3d-c4eb-4554-be1d-d4b07c4afe8b.png)

**Comment by orgtre** [comment]
This is a very useful discovery and should probably be communicated somewhere else too! 

To make the first step slightly easier one could use my qpdf transient wrapper [qpdf.el](https://github.com/orgtre/qpdf.el). In fact the whole procedure can be automated with the following command:
```elisp
(defun my-fix-pdf-selection ()
  "Replace pdf with one where selection shows transparently."
  (interactive)
  (unless (equal (file-name-extension (buffer-file-name)) "pdf")
    (error "Buffer should visit a pdf file."))
  (unless (or (equal major-mode 'pdf-view-mode))
    (pdf-view-mode))
  ;; save file in QDF-mode
  (qpdf-run (list
	     (concat "--infile="
		     (buffer-file-name))
	     "--qdf --object-streams=disable"
	     "--replace-input"))
  ;; do replacements
  (text-mode)
  (read-only-mode -1)
  (while (re-search-forward "3 Tr" nil t)
    (replace-match "7 Tr" nil nil))
  (save-buffer)
  (pdf-view-mode)
  ;; linearize file to disable QDF-mode
  (qpdf-run (list
	     (concat "--infile="
		     (buffer-file-name))
	     "--linearize"
	     "--replace-input"))
  (save-buffer)
  (pdf-view-mode))
```
This still depends on qpdf.el but with a few extra lines of code one could avoid that. Note that this overwrite the pdf file visited in the buffer from which it is run! To avoid this replace the first `"--replace-input"` with `(concat "--outfile=" (file-truename (read-file-name "Outfile: ")))`. The second call to qpdf to linearize isn't strictly necessary but it undoes the QDF-mode and can substantially reduce the filesize.

**Comment by workcomplete** [comment]
Great, thanks for sharing your package and function.

Reopening as it could be helpful if your function was added to the wiki, or at the very least if this could be added to known issues. 

One small issue with your function above is that when tested it only worked on file names that do not contain spaces.

**Comment by orgtre** [comment]
Thanks for testing. I fixed the issue with spaces in file names, plus added the problem and workaround to known problems in the README. If I understand correctly, once the pull request is merged, [pdftools.wiki](https://pdftools.wiki) (which just mirrors the README) will rebuild automatically to reflect the changes. I tried to work around the [qpdf.el](https://github.com/orgtre/qpdf.el) dependency, but to make it behave correctly in all cases, basically the whole `qpdf--default-run-after-function` is needed, so I decided to keep the dependency after all.

**Comment by workcomplete** [comment]
Thanks again for your work on this! FYI the `my-fix-pdf-selection` function recently stopped working on my setup, the first part of the function appears to be inserting a trailing `''`. 

Example:

```
call: qpdf '/home/user/Zotero/storage/Z4JZ82SF/some_file.pdf' --qdf --object-streams=disable --replace-input ''
```
When I copy this into a terminal window and remove the trailing `''` the command runs as expected.

**Comment by vedang** [comment]
@workcomplete : ^ This might be a bug introduced in the latest commit on `qpdf`: https://github.com/orgtre/qpdf.el/commit/6debdceef657691d678dcb24f91837f195d1b6b6 cc: @orgtre . I'm saying this after a cursory look at the linked commit, which adds a quotes around `outfile` (and in the command above there is no `outfile` argument). I might be wrong, I haven't actually tested the code.

I am going ahead and merging the PR which explains this as a known problem (which will also update the wiki). 

Thank you!

**Comment by vedang** [comment]
closed via vedang/pdf-tools@aec8ecd0066ba2cbfefed1330a71deada89f5b5a

**Comment by orgtre** [comment]
I merged vedang's fix. Thanks both of you!
